pkg/querier/blocks_store_querable_streaming.go

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storegateway/storegatewaypb"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type StreamingSeries struct {
+	Labels []mimirpb.LabelAdapter
+	Source StreamingSeriesSource
+}
+
+// StreamingSeriesSource holds the relationship between a stream of chunks from a SeriesChunksStreamReader
+// and the expected position of a series' chunks in that stream.
+type StreamingSeriesSource struct {
+	StreamReader *SeriesChunksStreamReader
+	SeriesIndex  uint64
+}
+
+// SeriesChunksStreamReader is responsible for managing the streaming of chunks from an ingester and buffering
+// chunks in memory until they are consumed by the PromQL engine.
+type SeriesChunksStreamReader struct {
+	client              storegatewaypb.StoreGateway_SeriesClient
+	expectedSeriesCount int
+	queryLimiter        *limiter.QueryLimiter
+	stats               *stats.Stats
+	log                 log.Logger
+
+	seriesCunksChan chan *storepb.StreamSeriesChunks
```

Comment:
For streaming from ingesters, I found the overhead of posting to a channel for each series to be substantial. Posting batches of series to the channel made a big difference.

It might be different given the relative amount of data we expect to receive from store-gateways, but might be something to experiment with if this turns out to be an issue here too.

Reply:
I have switched to sending batches of series when sending chunks, so this should be good now